examples/experimental/group_discussion_agents/group_discussion_agents.py

```diff
@@ -0,0 +1,169 @@
+    action_type: ActionType = Field(
+        description="whether to speak at this turn or choose to not do anything"
+    )
+    argument: str = Field(
+        description="the utterance if choose to speak, the expression or gesture if choose non-verbal communication, or the physical action if choose action"
+    )
+
+    def to_natural_language(self) -> str:
+        match self.action_type:
+            case "none":
+                return "did nothing"
+            case "speak":
+                return f'said: "{self.argument}"'
+            case "non-verbal communication":
+                return f"[{self.action_type}] {self.argument}"
+            case "action":
+                return f"[{self.action_type}] {self.argument}"
+            case "leave":
+                return "left the conversation"
+
+
+def _format_message_history(message_history: list[tuple[str, str]]) -> str:
+    return "\n".join(
+        (f"{speaker} said {message}") for speaker, message in message_history
+    )
+
+
+@NodeFactory.register("llm_agent")
+class LLMAgent(BaseAgent[AgentAction | Tick, AgentAction]):
+    def __init__(
+        self,
+        input_text_channels: list[str],
+        input_tick_channel: str,
+        output_channel: str,
+        query_interval: int,
+        agent_name: str,
+        goal: str,
+        model_name: str,
+        redis_url: str,
+    ):
+        super().__init__(
+            [
+                (input_text_channel, AgentAction)
+                for input_text_channel in input_text_channels
+            ]
+            + [
+                (input_tick_channel, Tick),
+            ],
+            [(output_channel, AgentAction)],
+            redis_url,
+        )
+        self.output_channel = output_channel
+        self.query_interval = query_interval
+        self.count_ticks = 0
+        self.message_history: list[tuple[str, str]] = []
+        self.name = agent_name
+        self.model_name = model_name
+        self.goal = goal
+
+    async def send(self, message: AgentAction) -> None:
+        if message.action_type == "speak":
+            await self.r.publish(
+                self.output_channel,
+                Message[AgentAction](data=message).model_dump_json(),
+            )
+
+    async def aact(self, message: AgentAction | Tick) -> AgentAction:
+        match message:
+            case Tick():
+                self.count_ticks += 1
+                if self.count_ticks % self.query_interval == 0:
+                    agent_action: str = await agenerate(
+                        model_name=self.model_name,
+                        template="Imagine that you are a friend of the other persons. Here is the "
+                        "conversation between you and them.\n"
+                        "You are {agent_name} in the conversation.\n"
+                        "{message_history}\n"
+                        "and you plan to {goal}.\n"
+                        "You can choose to interrupt the other person "
```

Comment:
How the interruption happens?

Reply:
If they output something they would interrupt the other agent, but if they don't output anything, the other agent can go on with their previous sentence.